From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Thu, 9 Dec 2021 13:27:56 -0500
Subject: [PATCH] Revert "Sentry Logging Support"

This reverts commit 74c0b90cc6569a2b367c63482bbc79688ae6fb2d.

diff --git a/build.gradle.kts b/build.gradle.kts
index c59e4260e474f4c1e35bd24a4d4f90f4ee980ff8..3f6898e43ea4578fdb9ee06301148dee5650623a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -64,8 +64,6 @@ dependencies {
     } // Airplane - more config
     implementation("com.github.technove:Flare:2c4a2114a0") // Airplane - flare
 
-    implementation("io.sentry:sentry:5.4.0") // Airplane
-
     testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
     testImplementation("junit:junit:4.13.1")
     testImplementation("org.hamcrest:hamcrest-library:1.3")
diff --git a/src/main/java/gg/airplane/AirplaneConfig.java b/src/main/java/gg/airplane/AirplaneConfig.java
index f34327d578ea8acfd4bb8f6f933af5d4f0f49c90..2b654e6adfe331a5fa68c1aea7d6d6cea1f3567d 100644
--- a/src/main/java/gg/airplane/AirplaneConfig.java
+++ b/src/main/java/gg/airplane/AirplaneConfig.java
@@ -2,7 +2,6 @@ package gg.airplane;
 
 import co.technove.air.AIR;
 import co.technove.air.ValueType;
-import io.sentry.Sentry;
 import net.minecraft.core.Registry;
 import gg.airplane.flare.FlareCommand;
 import net.minecraft.server.MinecraftServer;
@@ -254,33 +253,5 @@ public class AirplaneConfig {
         config.setComment("misc", "Settings for things that don't belong elsewhere");
     }
 
-    public static boolean usingSentry = false;
-    public static Level sentryLevel;
-
-    private static void sentrySetup() {
-        if (Sentry.isEnabled()) {
-            Sentry.endSession();
-            Sentry.close();
-        }
-
-        String sentryDsn = getString("sentry.dsn", "");
-        if (sentryDsn.isEmpty()) {
-            usingSentry = false;
-            Sentry.close();
-            return;
-        }
-
-        sentryLevel = Level.valueOf(getString("sentry.log-level", Level.WARN.name()));
-
-        usingSentry = true;
-        Sentry.init(sentryDsn);
-    }
-
-    public static void reportException(Throwable e) {
-        if (usingSentry) {
-            Sentry.captureException(e);
-        }
-    }
-
 
 }
diff --git a/src/main/java/gg/airplane/AirplaneSentryAppender.java b/src/main/java/gg/airplane/AirplaneSentryAppender.java
deleted file mode 100644
index eb83e44fc384e1b17291066d58512803834bf4ed..0000000000000000000000000000000000000000
--- a/src/main/java/gg/airplane/AirplaneSentryAppender.java
+++ /dev/null
@@ -1,35 +0,0 @@
-package gg.airplane;
-
-import io.sentry.Sentry;
-import org.apache.logging.log4j.core.Appender;
-import org.apache.logging.log4j.core.Core;
-import org.apache.logging.log4j.core.LogEvent;
-import org.apache.logging.log4j.core.appender.AbstractAppender;
-import org.apache.logging.log4j.core.config.plugins.Plugin;
-import org.apache.logging.log4j.core.config.plugins.PluginFactory;
-
-@Plugin(
-        name = "AirplaneSentry",
-        category = Core.CATEGORY_NAME,
-        elementType = Appender.ELEMENT_TYPE
-)
-public class AirplaneSentryAppender extends AbstractAppender {
-    protected AirplaneSentryAppender() {
-        super("AirplaneSentry", null, null);
-    }
-
-    @Override
-    public void append(LogEvent event) {
-        if (AirplaneConfig.usingSentry && event.getLevel().isMoreSpecificThan(AirplaneConfig.sentryLevel)) {
-            Throwable thrown = event.getThrown();
-            if (thrown != null) {
-                Sentry.captureException(thrown);
-            }
-        }
-    }
-
-    @PluginFactory
-    public static AirplaneSentryAppender createAppender() {
-        return new AirplaneSentryAppender();
-    }
-}
diff --git a/src/main/java/net/minecraft/CrashReport.java b/src/main/java/net/minecraft/CrashReport.java
index 1233c2d6b58a36a87e84cc73de7a71980431b9ca..b9a618eba86656289e20d051c3f0023ec93bb733 100644
--- a/src/main/java/net/minecraft/CrashReport.java
+++ b/src/main/java/net/minecraft/CrashReport.java
@@ -30,7 +30,6 @@ public class CrashReport {
     private final SystemReport systemReport = new SystemReport();
 
     public CrashReport(String message, Throwable cause) {
-        gg.airplane.AirplaneConfig.reportException(cause); // Airplane
         io.papermc.paper.util.StacktraceDeobfuscator.INSTANCE.deobfuscateThrowable(cause); // Paper
         this.title = message;
         this.exception = cause;
diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
index 77d71eaac32a2d33d7659134114c79415da64cd8..74ccc67e3c12dc5182602fb691ef3ddeb5b53280 100644
--- a/src/main/resources/log4j2.xml
+++ b/src/main/resources/log4j2.xml
@@ -42,7 +42,6 @@
             <AppenderRef ref="TerminalConsole" level="info"/>
             <AppenderRef ref="ServerGuiConsole" level="info"/>
         </Rewrite>
-        <AirplaneSentry />
     </Appenders>
     <Loggers>
         <Root level="info">
@@ -50,7 +49,6 @@
                 <MarkerFilter marker="NETWORK_PACKETS" onMatch="DENY" onMismatch="NEUTRAL" />
             </filters>
             <AppenderRef ref="Async"/>
-            <AppenderRef ref="AirplaneSentry"/>
         </Root>
     </Loggers>
 </Configuration>
