From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Sat, 2 Apr 2022 02:12:13 -0400
Subject: [PATCH] Fix issue with lecterns


diff --git a/src/main/java/net/minecraft/world/inventory/LecternMenu.java b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
index ff79925bc6437222f9ceb133e21bbc0600cc74ed..8674b35c12c87bd169d834801e71c1c5c71dc9bc 100644
--- a/src/main/java/net/minecraft/world/inventory/LecternMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
@@ -119,6 +119,11 @@ public class LecternMenu extends AbstractContainerMenu {
             }
         }
     }
+    
+    @Override
+    public ItemStack quickMoveStack(net.minecraft.world.entity.player.Player player, int index) {
+        return ItemStack.EMPTY;
+    }
 
     @Override
     public void setData(int id, int value) {
